[axonius][network] Add Axonius Network datastream

muskan-agarwal26 opened the pull request:
## Proposed commit message

The release includes network data stream and associated dashboard.

Axonius fields are mapped to their corresponding ECS fields where possible.

Test samples were derived from live data samples, which were subsequently
sanitized.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [x] I have verified that any added dashboard complies with Kibana's [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices) 

## How to test this PR locally

To test the axonius package:

- Clone integrations repo.
- Install elastic package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/axonius directory.
- Run the following command to run tests.

> elastic-package test
```
2025/12/21 16:36:24  INFO New version is available - v0.117.1. Download from: https://github.com/elastic/elastic-package/releases/tag/v0.117.1
Run asset tests for the package
2025/12/21 16:36:25  INFO License text found in "/root/GITHUB/integrations/LICENSE.txt" will be included in package
--- Test results for package: axonius - START ---
╭─────────┬─────────────┬───────────┬──────────────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE │ DATA STREAM │ TEST TYPE │ TEST NAME                                                        │ RESULT │ TIME ELAPSED │
├─────────┼─────────────┼───────────┼──────────────────────────────────────────────────────────────────┼────────┼──────────────┤
│ axonius │             │ asset     │ dashboard axonius-a47bc47d-5f69-473f-92a1-07a79cc71cf8 is loaded │ PASS   │      1.847µs │
│ axonius │ network     │ asset     │ index_template logs-axonius.network is loaded                    │ PASS   │        256ns │
│ axonius │ network     │ asset     │ ingest_pipeline logs-axonius.network-0.1.0 is loaded             │ PASS   │        275ns │
╰─────────┴─────────────┴───────────┴──────────────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: axonius - END   ---
Done
Run pipeline tests for the package
--- Test results for package: axonius - START ---
╭─────────┬─────────────┬───────────┬────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE │ DATA STREAM │ TEST TYPE │ TEST NAME                                          │ RESULT │ TIME ELAPSED │
├─────────┼─────────────┼───────────┼────────────────────────────────────────────────────┼────────┼──────────────┤
│ axonius │ network     │ pipeline  │ (ingest pipeline warnings test-load-balancer.log)  │ PASS   │ 762.895972ms │
│ axonius │ network     │ pipeline  │ (ingest pipeline warnings test-network-device.log) │ PASS   │  693.13953ms │
│ axonius │ network     │ pipeline  │ test-load-balancer.log                             │ PASS   │ 233.874266ms │
│ axonius │ network     │ pipeline  │ test-network-device.log                            │ PASS   │ 347.384569ms │
╰─────────┴─────────────┴───────────┴────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: axonius - END   ---
Done
Run policy tests for the package
--- Test results for package: axonius - START ---
No test results
--- Test results for package: axonius - END   ---
Done
Run static tests for the package
--- Test results for package: axonius - START ---
╭─────────┬─────────────┬───────────┬──────────────────────────┬────────┬──────────────╮
│ PACKAGE │ DATA STREAM │ TEST TYPE │ TEST NAME                │ RESULT │ TIME ELAPSED │
├─────────┼─────────────┼───────────┼──────────────────────────┼────────┼──────────────┤
│ axonius │ network     │ static    │ Verify sample_event.json │ PASS   │  249.74697ms │
╰─────────┴─────────────┴───────────┴──────────────────────────┴────────┴──────────────╯
--- Test results for package: axonius - END   ---
Done
Run system tests for the package
2025/12/21 16:36:34  INFO Installing package...
2025/12/21 16:36:35  INFO License text found in "/root/GITHUB/integrations/LICENSE.txt" will be included in package
2025/12/21 16:36:49  INFO Running test for data_stream "network" with configuration 'default'
2025/12/21 16:36:57  INFO Setting up independent Elastic Agent...
2025/12/21 16:37:08  INFO Setting up service...
2025/12/21 16:37:33  INFO Validating test case...
2025/12/21 16:37:35  INFO Tearing down service...
2025/12/21 16:37:36  INFO Write container logs to file: /root/GITHUB/integrations/build/container-logs/axonius-1766315256251921752.log
2025/12/21 16:37:38  INFO Tearing down agent...
2025/12/21 16:37:39  INFO Write container logs to file: /root/GITHUB/integrations/build/container-logs/elastic-agent-1766315259135394865.log
2025/12/21 16:39:21  INFO Uninstalling package...
--- Test results for package: axonius - START ---
╭─────────┬─────────────┬───────────┬───────────┬────────┬───────────────╮
│ PACKAGE │ DATA STREAM │ TEST TYPE │ TEST NAME │ RESULT │  TIME ELAPSED │
├─────────┼─────────────┼───────────┼───────────┼────────┼───────────────┤
│ axonius │ network     │ system    │ default   │ PASS   │ 46.195212588s │
╰─────────┴─────────────┴───────────┴───────────┴────────┴───────────────╯
--- Test results for package: axonius - END   ---
Done
``` 

## Related issues

- Relates https://github.com/elastic/integrations/issues/15034
- Relates https://github.com/elastic/elasticsearch/pull/139512

## Screenshots
<img width="1919" height="913" alt="image (5)" src="https://github.com/user-attachments/assets/c254d523-7517-4137-a726-84e7c77f34fa" />
<img width="1919" height="916" alt="image (6)" src="https://github.com/user-attachments/assets/9abed0f7-58f5-4b4e-aca9-474d6de7b50f" />

muskan-crest committed:
Added network datastream in axonius.

muskan-crest committed:
Update changelog entry.

muskan-crest committed:
Added few mappings of missing fields

muskan-crest committed:
Added transform and updated logo

muskan-crest committed:
Merge branch 'feature/axonius-0.1.0' of github.com:muskan-agarwal26/integrations into datastream-network

muskan-crest committed:
Took sync from latest feature/axonius-0.1.0 branch.

muskan-crest committed:
Added filters on visual level.

efd6 commented on packages/axonius/data_stream/network/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,370 @@
+  - script:
+      tag: script_rename_event_data_fields
+      lang: painless
+      description: Renames all event.data.* fields to root level (removes event.data prefix). Prefixes conflicting keys with 'data_' to avoid overwriting root-level fields.
```
```suggestion
      description: Renames all axonius.network.event.data.* fields to root level (removes event.data prefix). Prefixes conflicting keys with 'data_' to avoid overwriting custom data-stream-root-level fields.
```

efd6 commented on packages/axonius/data_stream/network/elasticsearch/ingest_pipeline/pipeline-nat-rule.yml:
```diff
@@ -0,0 +1,224 @@
+      on_failure:
+        - append:
+            field: error.message
+            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.on_failure_pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
```
Same as https://github.com/elastic/integrations/pull/16657/changes#r3000751036 here.

efd6 commented on packages/axonius/data_stream/network/elasticsearch/ingest_pipeline/pipeline-network-device.yml:
```diff
@@ -0,0 +1,2272 @@
+      on_failure:
+        - append:
+            field: error.message
+            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.on_failure_pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
```
Also here and below.

efd6 commented on packages/axonius/data_stream/network/elasticsearch/ingest_pipeline/pipeline-network.yml:
```diff
@@ -0,0 +1,204 @@
+      on_failure:
+        - append:
+            field: error.message
+            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.on_failure_pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
```
Also here and below.

efd6 commented on packages/axonius/kibana/dashboard/axonius-a47bc47d-5f69-473f-92a1-07a79cc71cf8.json:
Do we want a link panel?

efd6 commented on packages/axonius/data_stream/network/elasticsearch/ingest_pipeline/pipeline-network-device.yml:
It is not possible to include an `if` condition in a processor child of a `foreach` processor, elastic/elasticsearch#60470.

efd6 replied:
Sure

muskan-crest committed:
resolved bot and @efd6 comments.

muskan-crest committed:
Updated if cond in default.yml as suggested by bot.

kcreddy commented:
/test

kcreddy commented on packages/axonius/manifest.yml:
I think we need to add more categories based on the data streams. Please check [this enum list](https://github.com/elastic/package-spec/blob/main/spec/integration/manifest.spec.yml#L21) and add accordingly for all data streams that are added until now.

kcreddy commented on packages/axonius/elasticsearch/transform/latest_network/transform.yml:
```diff
@@ -0,0 +1,37 @@
+    - axonius.network.transform_unique_id
+  sort: '@timestamp'
+description: >-
+  Latest networks from Axonius. As networks get updated, this transform stores only the latest state of each network asset inside the destination index. Thus the transform's destination index contains only the latest state of the network asset.
```
The retention approach will eventually work, but remain stale until retention gets triggered.
I am looking at the `config.yml` used for system tests. Since we are splitting on `assets[].specific_data[]`, have you tried a unique id that is a combination:
- `assets[].internal_axon_id` + `assets[].specific_data[].id` OR
- `assets[].internal_axon_id` + `assets[].specific_data[].id_raw`
That may seem to give us a unique identifier for each network asset.

kcreddy replied:
It should be in the `Overview`

muskan-crest committed:
resolved comments provided by @kcreddy.
1. Added categories in manifest for each datastream.
2. changed ilm policy retention from 90d to 30d.
3. Added remove proc inside convert.

kcreddy reviewed (commented):
Update README and Dashboard overview to reflect limitations discussed in https://github.com/elastic/integrations/pull/16657/changes#r3056389784

muskan-crest committed:
Added note about transform retention in readme and dashboard.

muskan-agarwal26 commented:
I have added a note in both as discussed. @kcreddy

kcreddy commented on packages/axonius/_dev/build/docs/README.md:
```diff
@@ -129,6 +138,9 @@ Destinations indices are aliased to `logs-axonius_latest.<data_stream_name>`.
 | `logs-axonius.ticket-*`              | `logs-axonius_latest.dest_ticket-*`                | `logs-axonius_latest.ticket`
+| `logs-axonius.network-*`              | `logs-axonius_latest.dest_network-*`                | `logs-axonius_latest.network`
+
+**Note:** Assets deleted from Axonius may reappear in a future discovery cycle if they are still present in connected data sources and get re-detected. Because the exact duration for which a deleted asset may remain dormant before being rediscovered is unknown, the transform retention period is set to **90 days** to reduce the risk of data loss for such assets.
```
I think we are not capturing the real user-facing consequence. You should also add:
`This means deleted assets will continue to appear in dashboards for up to 90 days after deletion.`

kcreddy replied:
Also add the same to the dashboard overview.

kcreddy commented on packages/axonius/_dev/build/docs/README.md:
```diff
@@ -129,6 +138,9 @@ Destinations indices are aliased to `logs-axonius_latest.<data_stream_name>`.
 | `logs-axonius.ticket-*`              | `logs-axonius_latest.dest_ticket-*`                | `logs-axonius_latest.ticket`
+| `logs-axonius.network-*`              | `logs-axonius_latest.dest_network-*`                | `logs-axonius_latest.network`
+
+**Note:** Assets deleted from Axonius may reappear in a future discovery cycle if they are still present in connected data sources and get re-detected. Because the exact duration for which a deleted asset may remain dormant before being rediscovered is unknown, the transform retention period is set to **90 days** to reduce the risk of data loss for such assets.
```
The users should also understand the difference between this data stream and other datastreams such as `user`, `gateway`, etc. (which has unique entity id which actually contains latest state of the entity).

Add:
`The network destination index is a content-based deduplicated view, not an entity-level latest-state view like the other data streams.`

kcreddy commented on packages/axonius/elasticsearch/transform/latest_network/transform.yml:
```diff
@@ -0,0 +1,37 @@
+    - axonius.network.transform_unique_id
+  sort: '@timestamp'
+description: >-
+  Latest networks from Axonius. As networks get updated, this transform stores only the latest state of each network asset inside the destination index. Thus the transform's destination index contains only the latest state of the network asset.
```
The description is misleading: `As networks get updated, this transform stores only the latest state of each network asset`
But you said as per the vendor, network assets are never updated, but only created and deleted.
The description should accurately reflect what the transform does now: deduplicates identical documents from repeated API fetches.

kcreddy commented on packages/axonius/elasticsearch/transform/latest_network/transform.yml:
```diff
@@ -0,0 +1,37 @@
+  aliases:
+    - alias: 'logs-axonius_latest.network'
+      move_on_creation: true
+latest:
+  unique_key:
+    - event.dataset
+    - axonius.network.transform_unique_id
```
Add code comments explaining the choice of unique key.

- The Axonius API provides no stable unique identifier for network assets
- `transform_unique_id` is a fingerprint of `event.original` (the full document)
- Per vendor: network assets cannot be updated, only created and deleted
- Known fragility: if the API response format changes (field ordering, new fields, etc.), the fingerprint will differ, creating duplicate entries in the destination index

kcreddy commented on packages/axonius/data_stream/network/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,390 @@
+      if: ctx.event?.original != null
+  - json:
+      field: event.original
+      tag: json_event_original
+      target_field: axonius.network
+  - fingerprint:
+      fields:
+        - event.original
+      tag: fingerprint_event_original
+      target_field: axonius.network.transform_unique_id
+      ignore_missing: true
```
Here, there should be explanation to why we fingerprint full document.
Add `description` or code comments noting that it exists because no stable entity ID is available from the API.

muskan-crest committed:
Resolved readme and dashboard notes comments provided by @kcreddy.

kcreddy reviewed (approved):
LGTM, please wait for @efd6 approval.

efd6 commented:
My concerns have not been addressed.

muskan-crest committed:
Reoslved cel comments provided by @efd6.

muskan-agarwal26 commented:
@efd6 , I have addressed your comments, please have a look.
